src/wp-admin/includes/plugin.php

```diff
@@ -1414,6 +1414,40 @@ function add_comments_page( $page_title, $menu_title, $capability, $menu_slug, $
 	return add_submenu_page( 'edit-comments.php', $page_title, $menu_title, $capability, $menu_slug, $function );
 }
 
+/**
+ * Add submenu page to the Security main menu.
+ *
+ * The function which is hooked in to handle the output of the page must check
+ * that the user has at least 'manage_options' capability.
+ *
+ * @param string   $page_title The text to be displayed in the title tags of the page when the menu is selected.
+ * @param string   $menu_title The text to be used for the menu.
+ * @param string   $menu_slug  The slug name to refer to this menu by; must match an active plugin slug.
+ * @param callable $function   The function to be called to output the content for this page.
+ * @return false|string The resulting page's hook_suffix, or false if the user does not have the capability required.
+ */
+function add_security_page( $page_title, $menu_title, $menu_slug, $function = '' ) {
+    if ( WP_DEBUG ) {
+        foreach ( get_option( 'active_plugins' ) as $path ) {
+            $parts = explode( '/', $path );
+            if ( $menu_slug === $parts[0] ) {
+                add_filter(
+                    'plugin_action_links_' . $path,
+                    function ( $links ) use ( $menu_slug ) {
+                        array_unshift( $links, sprintf( '<a href="%s?page=%s" title="%s"><span class="dashicons-shield" style="font: 16px dashicons; vertical-align: text-bottom;"></span></a>', admin_url( 'security.php' ), $menu_slug, __( 'Security' ) ) );
+                        return $links;
+                    }
+                );
+                return add_submenu_page( 'security.php', $page_title, $menu_title, 'manage_options', $menu_slug, $function );
+            }
+        }
+
+        _doing_it_wrong(__METHOD__, '$menu_slug must match an active plugin slug; '.$menu_slug, '1.1.0');
```

Comment:
1. I think this may be too much for a v1, because we haven't seen how people are going to use the new screen yet.
2. Having this only enabled on `WP_DEBUG` is not a good user experience, so let's enable it always, or take it out. Some (few) users will see it, and others will not. This is `O(number of active plugins)`, and it will only ever run in the admin dashboard, which is not a big deal.
3. If this is staying then please post a screenshot to the PR so we can see how it looks. This screenshot should then end up on a documentation page so that developers and users have more details about what to expect.

Reply:
The idea is that developers will see this - so maybe `wp_die()` is a better choice - if they try to use a menu slug that doesn't match their plugin slug; an end user should never see it. If there's a better way to do that I'll change it.
The idea of having the slugs match is that later on we can reason about plugins and security pages without having to grovel through `$submenu`. Maybe that's too ambitious ;-)

Reply:
After seeing a screenshot over on the forum thread I like this, it is like a little badge of honor for plugins that do things according to the recommendations.

>The idea of having the slugs match is that later on we can reason about plugins and security pages without having to grovel through `$submenu`. Maybe that's too ambitious ;-)

Trying to bolt this on after the fact is even more ambitious. I see how this would be useful so let's get it in there from the start.

>if they try to use a menu slug that doesn't match their plugin slug; an end user should never see it

I'm fine with this too, but the only way to really guarantee it is to make the page actually not show up ever. So let's drop the `WP_DEBUG` conditional and always enforce the check. I would say `_doing_it_wrong` combined with not actually registering the page is the way to go here.